### PR TITLE
Memento pattern: full and understandable implementation

### PR DESCRIPTION
**Pull request by KIVagant**
- Caretaker should save history for normal pattern implementation.
- Importent notes in code of Originator.
- Readme with full pattern philosophy.
- Tests and example of usage inside test.

Please, check my English especially inside Readme. Thank you.

### CODE REVIEW

File: Behavioral/Memento/README.rst [hunk 1]
```diff
@@ -4,26 +4,43 @@
...
-asks the originator for a memento object. Then it does whatever
-operation (or sequence of operations) it was going to do. To roll back
-to the state before the operations, it returns the memento object to the
-originator. The memento object itself is an opaque object (one which the
-caretaker cannot, or should not, change). When using this pattern, care
-should be taken if the originator may change other objects or resources
-- the memento pattern operates on a single object.
+It provides the ability to restore an object to its previous state (undo
+via rollback) or to gain access to state of the object, without revealing
+its implementation (ie, the object is not required to have a functional
```
Comment (domnikl) [inline Behavioral/Memento/README.rst]: `i.e.`

File: Behavioral/Memento/README.rst [hunk 2]
```diff
@@ -4,26 +4,43 @@
...
-internal state. The caretaker is going to do something to the
-originator, but wants to be able to undo the change. The caretaker first
-asks the originator for a memento object. Then it does whatever
-operation (or sequence of operations) it was going to do. To roll back
-to the state before the operations, it returns the memento object to the
-originator. The memento object itself is an opaque object (one which the
-caretaker cannot, or should not, change). When using this pattern, care
-should be taken if the originator may change other objects or resources
-- the memento pattern operates on a single object.
+It provides the ability to restore an object to its previous state (undo
```
Comment (domnikl) [inline Behavioral/Memento/README.rst]: Please check all `its` and correct to `it's`

File: Behavioral/Memento/README.rst [hunk 3]
```diff
@@ -4,26 +4,43 @@
...
+via rollback) or to gain access to state of the object, without revealing
+its implementation (ie, the object is not required to have a functional
+for return the current state).
+
+The memento pattern is implemented with three objects: the Originator, a
+Caretaker and a Memento.
+
+Memento – an object that *contains a concrete unique snapshot of state* of
+any object or resource: string, number, array, an instance of class and so on.
+The uniqueness in this case  not imply the prohibition existence of similar
```
Comment (domnikl) [inline Behavioral/Memento/README.rst]: in this case *does* not imply

**Comment by domnikl** [comment]
Apart from these errors in the text I commented on, it looks very good. Sorry for let you wait so long for feedback on this PR!

**Comment by KIVagant** [comment]
@domnikl, thank you. I have corrected grammatical errors. Now it's looks better?

**Comment by domnikl** [comment]
@KIVagant yes looks good now!

**Pull request merged**